File: scripts/python/src/tools/dotfile/secret/variables.py
import json
import os
import re

from tools.core.process import capture
from tools.dotfile.secret.identity import identity_path, sops_env

VARS = "vars.enc.yaml"
OPEN = "open."
PLACEHOLDER = re.compile(r"\{\{\s*([A-Za-z0-9_][A-Za-z0-9_.-]*)\s*\}\}")


class Vars:
    def __init__(self, values, ok, note):
        self.values = values
        self.ok = ok
        self.note = note


def vars_file(ctx):
    return os.path.join(ctx.root, VARS)


def scalar(value):
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def flatten(node, prefix, out):
    for key, value in node.items():
        name = f"{prefix}.{key}" if prefix else key
        if isinstance(value, dict):
            problem = flatten(value, name, out)
            if problem:
                return problem
        elif isinstance(value, list):
            return f"'{name}' is a list; a var must be a single value"
        elif value is None:
            return f"'{name}' has no value"
        else:
            out[name] = scalar(value)
    return ""


def load(ctx):
    path = vars_file(ctx)
    if not os.path.isfile(path):
        return Vars({}, True, "")
    if not os.path.isfile(identity_path(ctx)):
        return Vars({}, False, f"{VARS} needs an age identity to read")
    result = capture(["sops", "-d", "--output-type", "json", path], env=sops_env(ctx))
    if result.returncode != 0:
        return Vars({}, False, f"{VARS} did not decrypt on this machine")
    try:
        raw = json.loads(result.stdout)
    except ValueError:
        return Vars({}, False, f"{VARS} did not decrypt to an object")
    if not isinstance(raw, dict):
        return Vars({}, False, f"{VARS} must hold a mapping")
    values = {}
    problem = flatten(raw, "", values)
    if problem:
        return Vars({}, False, problem)
    return Vars(values, True, "")


def references(text):
    return sorted({match.group(1) for match in PLACEHOLDER.finditer(text)})


def render(text, values):
    missing = []

    def substitute(match):
        name = match.group(1)
        if name in values:
            return values[name]
        missing.append(name)
        return match.group(0)

    rendered = PLACEHOLDER.sub(substitute, text)
    return rendered, sorted(set(missing))

File: scripts/rust/crates/dotfile/src/secret/variables.rs
use crate::context::Context;
use serde_json::Value;
use std::collections::{BTreeMap, BTreeSet};

pub struct Variables {
    pub values: BTreeMap<String, String>,
    pub ok: bool,
    pub note: String,
}

pub fn load_variables(context: &Context) -> Variables {
    let source = context.root.join("vars.enc.yaml");
    if !source.is_file() {
        return Variables {
            values: BTreeMap::new(),
            ok: true,
            note: String::new(),
        };
    }
    if !super::vault::identity_path(context).is_file() {
        return Variables {
            values: BTreeMap::new(),
            ok: false,
            note: "vars.enc.yaml needs an age identity to read".to_string(),
        };
    }
    let output = match super::sops::decrypt(context, &source, None, true) {
        Ok(output) => output,
        Err(note) => {
            return Variables {
                values: BTreeMap::new(),
                ok: false,
                note,
            };
        }
    };
    let Ok(Value::Object(document)) = serde_json::from_slice::<Value>(&output) else {
        return Variables {
            values: BTreeMap::new(),
            ok: false,
            note: "vars.enc.yaml must hold a mapping".to_string(),
        };
    };
    let mut values = BTreeMap::new();
    match flatten_variables(&document, "", &mut values) {
        Ok(()) => Variables {
            values,
            ok: true,
            note: String::new(),
        },
        Err(note) => Variables {
            values: BTreeMap::new(),
            ok: false,
            note,
        },
    }
}

pub fn flatten_variables(
    values: &serde_json::Map<String, Value>,
    prefix: &str,
    output: &mut BTreeMap<String, String>,
) -> Result<(), String> {
    for (key, value) in values {
        let name = if prefix.is_empty() {
            key.clone()
        } else {
            format!("{prefix}.{key}")
        };
        match value {
            Value::Object(children) => flatten_variables(children, &name, output)?,
            Value::Array(_) => {
                return Err(format!("'{name}' is a list; a var must be a single value"));
            }
            Value::Null => return Err(format!("'{name}' has no value")),
            Value::Bool(value) => {
                output.insert(name, value.to_string());
            }
            Value::String(value) => {
                output.insert(name, value.clone());
            }
            Value::Number(value) => {
                output.insert(name, value.to_string());
            }
        }
    }
    Ok(())
}

static PLACEHOLDER: std::sync::LazyLock<regex::Regex> = std::sync::LazyLock::new(|| {
    regex::Regex::new(r"\{\{\s*([A-Za-z0-9_][A-Za-z0-9_.-]*)\s*\}\}")
        .expect("valid placeholder pattern")
});

pub fn references(template: &str) -> Vec<String> {
    PLACEHOLDER
        .captures_iter(template)
        .map(|capture| capture[1].to_string())
        .collect::<BTreeSet<_>>()
        .into_iter()
        .collect()
}

pub fn render_template(template: &str, values: &BTreeMap<String, String>) -> (String, Vec<String>) {
    let mut missing = BTreeSet::new();
    let output = PLACEHOLDER.replace_all(template, |capture: &regex::Captures<'_>| {
        if let Some(value) = values.get(&capture[1]) {
            value.clone()
        } else {
            missing.insert(capture[1].to_string());
            capture[0].to_string()
        }
    });
    (output.into_owned(), missing.into_iter().collect())
}

impl Drop for Variables {
    fn drop(&mut self) {
        use zeroize::Zeroize;
        for value in self.values.values_mut() {
            value.zeroize();
        }
    }
}
